tests/versioning_3_test.go

```diff
@@ -212,6 +218,123 @@ func (s *Versioning3Suite) testWorkflowWithPinnedOverride(sticky bool) {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testQueryWithPinnedOverride(false)
+		},
+	)
+}
+
+func (s *Versioning3Suite) TestQueryWithPinnedOverride_Sticky() {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testQueryWithPinnedOverride(true)
+		},
+	)
+}
+
+func (s *Versioning3Suite) testQueryWithPinnedOverride(sticky bool) {
+	tv := testvars.New(s)
+	handlingQueries := true
+
+	if sticky {
+		s.warmUpSticky(tv, handlingQueries)
+	}
+
+	wftCompleted := make(chan interface{})
+	s.pollWftAndHandle(tv, false, wftCompleted,
+		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondWorkflowTaskCompletedRequest, error) {
+			s.NotNil(task)
+			return respondEmptyWft(tv, sticky, vbUnpinned), nil
+		})
+
+	override := makePinnedOverride(tv.Deployment())
+	we := s.startWorkflow(tv, override)
+
+	<-wftCompleted
+	s.verifyWorkflowVersioning(tv, vbUnpinned, tv.Deployment(), override, nil)
+	if sticky {
+		s.verifyWorkflowStickyQueue(we, tv.StickyTaskQueue())
+	}
+
+	s.pollAndQueryWorkflow(tv, sticky)
+}
+
+func (s *Versioning3Suite) TestUnpinnedQuery_NoSticky() {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testUnpinnedQuery(false)
+		},
+	)
+}
+
+func (s *Versioning3Suite) TestUnpinnedQuery_Sticky() {
+	s.RunTestWithMatchingBehavior(
+		func() {
+			s.testUnpinnedQuery(true)
+		},
+	)
+}
+
+func (s *Versioning3Suite) testUnpinnedQuery(sticky bool) {
+	tv := testvars.New(s)
+	tvB := tv.WithBuildId("B")
+	d := tv.Deployment()
+	handlingQueries := true
+	if sticky {
+		s.warmUpSticky(tv, handlingQueries)
+	}
+
+	wftCompleted := make(chan interface{})
+	s.pollWftAndHandle(tv, false, wftCompleted,
+		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondWorkflowTaskCompletedRequest, error) {
+			s.NotNil(task)
+			s.verifyWorkflowVersioning(tv, vbUnspecified, nil, nil, transitionTo(d))
+			return respondEmptyWft(tv, sticky, vbUnpinned), nil
+		})
+
+	s.setCurrentDeployment(d)
+	s.waitForDeploymentDataPropagation(tv, tqTypeWf)
+
+	we := s.startWorkflow(tv, nil)
+
+	<-wftCompleted
+	s.verifyWorkflowVersioning(tv, vbUnpinned, d, nil, nil)
+	if sticky {
+		s.verifyWorkflowStickyQueue(we, tv.StickyTaskQueue())
+	}
+
+	go s.idlePollWorkflow(tvB, sticky, ver3MinPollTime, "new deployment should not receive query", handlingQueries)
+	s.pollAndQueryWorkflow(tv, sticky)
+
+	// redirect query to new deployment
+	s.updateTaskQueueDeploymentData(tvB, 0, tqTypeWf, tqTypeAct)
+
+	go s.idlePollWorkflow(tv, sticky, ver3MinPollTime, "old deployment should not receive query", handlingQueries)
+	s.pollAndQueryWorkflow(tvB, sticky)
+
+}
+
+func (s *Versioning3Suite) pollAndQueryWorkflow(
+	tv *testvars.TestVars,
+	sticky bool,
+) {
+	queryResultCh := make(chan interface{})
+	s.pollWftAndHandleQueries(tv, sticky, queryResultCh,
+		func(task *workflowservice.PollWorkflowTaskQueueResponse) (*workflowservice.RespondQueryTaskCompletedRequest, error) {
+			s.NotNil(task)
+			s.NotNil(task.Query)
```

Comment:
As mentioned elsewhere, I think the taskpoller should ensure that before invoking this handler, ie these lines should not be necessary.